File: backend/internal/context/task_logger.go
// Package context provides utilities for managing task logs and context operations.
package context

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// BaseDir specifies the root directory where backlog.md and codex_task_tracker.md reside.
// Tests can override this to use a temporary directory.
var BaseDir = ""

const (
	backlogFile = "backend/backlog.md"
	taskLogFile = "codex_task_tracker.md"
)

// resolveBaseDir returns the directory containing backlog.md and codex_task_tracker.md.
// If BaseDir is set (tests), it is used as-is. Otherwise the function walks up
// from the current working directory until it finds the expected files.
func resolveBaseDir() string {
	if BaseDir != "" {
		return BaseDir
	}
	dir, err := os.Getwd()
	if err != nil {
		return ""
	}
	for {
		if _, err := os.Stat(filepath.Join(dir, backlogFile)); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	return ""
}

// UpdateTaskTracker updates an existing task status in codex_task_tracker.md or
// appends a new entry when not present. When parentTaskName is provided, the
// entry will be prefixed with "parent – task" unless already prefixed. The
// variadic argument keeps backward compatibility for existing callers.
func UpdateTaskTracker(taskName, layers, status, assignedTo, notes string, parentTaskName ...string) error {
	parent := ""
	if len(parentTaskName) > 0 {
		parent = parentTaskName[0]
	}
	if parent != "" && !strings.HasPrefix(taskName, parent+" – ") {
		taskName = fmt.Sprintf("%s – %s", parent, taskName)
	}

	base := resolveBaseDir()
	fp := filepath.Join(base, taskLogFile)

	if hasDuplicateTask(taskName) {
		data, err := os.ReadFile(fp)
		if err != nil {
			return fmt.Errorf("could not read codex_task_tracker.md: %w", err)
		}

		lines := strings.Split(string(data), "\n")
		for i, line := range lines {
			fields := strings.Split(line, "|")
			if len(fields) < 5 {
				continue
			}
			name := strings.TrimSpace(fields[1])
			if name == taskName {
				created := time.Now().Format("2006-01-02")
				if len(fields) >= 7 {
					created = strings.TrimSpace(fields[6])
				}
				updated := time.Now().Format("2006-01-02")
				lines[i] = fmt.Sprintf("| %-25s | %-25s | %-13s | %-11s | %s | %-10s | %-10s |",
					taskName, layers, status, assignedTo, notes, created, updated)
				break
			}
		}

		output := strings.Join(lines, "\n")
		if err := os.WriteFile(fp, []byte(output), 0644); err != nil {
			return fmt.Errorf("could not write to codex_task_tracker.md: %w", err)
		}
	} else {
		file, err := os.OpenFile(fp, os.O_APPEND|os.O_WRONLY, 0644)
		if err != nil {
			return fmt.Errorf("could not open codex_task_tracker.md: %w", err)
		}

		now := time.Now().Format("2006-01-02")
		row := fmt.Sprintf("| %-25s | %-25s | %-13s | %-11s | %s | %-10s | %-10s |\n",
			taskName, layers, status, assignedTo, notes, now, now)

		if _, err := file.WriteString(row); err != nil {
			_ = file.Close()
			return fmt.Errorf("could not write to codex_task_tracker.md: %w", err)
		}

		if err := file.Close(); err != nil {
			fmt.Fprintf(os.Stderr, "file.Close error: %v\n", err)
		}
	}

	if err := CleanBacklog(); err != nil {
		return fmt.Errorf("clean backlog: %w", err)
	}

	return nil
}

// AppendSubtask is a helper that prefixes the subtask with the parent task name
// and delegates to UpdateTaskTracker.
func AppendSubtask(parentTask, subtask, layers, status, assignedTo, notes string) error {
	return UpdateTaskTracker(subtask, layers, status, assignedTo, notes, parentTask)
}

func hasDuplicateTask(taskName string) bool {
	base := resolveBaseDir()
	fp := filepath.Join(base, taskLogFile)
	file, err := os.Open(fp)
	if err != nil {
		return false // Safe fallback
	}
	defer func() {
		if err := file.Close(); err != nil {
			fmt.Fprintf(os.Stderr, "file.Close error: %v\n", err)
		}
	}()

	scanner := bufio.NewScanner(file)
	rgx := regexp.MustCompile(`\|\s*(.*?)\s*\|`)

	for scanner.Scan() {
		line := scanner.Text()
		if rgx.MatchString(line) {
			task := rgx.FindStringSubmatch(line)[1]
			if task == taskName {
				return true
			}
		}
	}
	return false
}

// CleanBacklog removes tasks from /backend/backlog.md once they are marked
// `✅ Done` in codex_task_tracker.md. It preserves all other content and is
// safe to run multiple times.
func CleanBacklog() error {
	base := resolveBaseDir()
	done, err := parseDoneTasks()
	if err != nil {
		return err
	}

	// include variations of done task names without optional prefixes
	doneNormalized := make(map[string]struct{}, len(done))
	for name := range done {
		doneNormalized[name] = struct{}{}
		if strings.Contains(name, "–") {
			parts := strings.SplitN(name, "–", 2)
			trimmed := strings.TrimSpace(parts[0])
			if trimmed != "" {
				doneNormalized[trimmed] = struct{}{}
			}
		}
	}

	backlogPath := filepath.Join(base, backlogFile)
	backlogBytes, err := os.ReadFile(backlogPath)
	if err != nil {
		return fmt.Errorf("could not read backlog.md: %w", err)
	}

	lines := strings.Split(string(backlogBytes), "\n")
	var result []string
	var skip bool

	headingRgx := regexp.MustCompile(`(?i)^.*Codex Task:`)

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if headingRgx.MatchString(trimmed) && !strings.HasPrefix(trimmed, "<!--") && !strings.HasPrefix(trimmed, "*") && !strings.HasPrefix(trimmed, "-") && !strings.ContainsAny(trimmed, "\"'") {
			idx := strings.Index(strings.ToLower(trimmed), "codex task:")
			if idx != -1 {
				taskName := strings.TrimSpace(trimmed[idx+len("Codex Task:"):])
				taskName = strings.Trim(taskName, "[]")
				if _, ok := doneNormalized[taskName]; ok {
					skip = true
					continue
				}
				skip = false
			}
		}

		if skip {
			continue
		}
		result = append(result, line)
	}

	cleaned := strings.Join(result, "\n")
	if strings.HasSuffix(string(backlogBytes), "\n") && !strings.HasSuffix(cleaned, "\n") {
		cleaned += "\n"
	}

	if cleaned == string(backlogBytes) {
		return nil
	}

	if err := os.WriteFile(backlogPath, []byte(cleaned), 0644); err != nil {
		return fmt.Errorf("could not write backlog.md: %w", err)
	}
	return nil
}

func parseDoneTasks() (map[string]struct{}, error) {
	base := resolveBaseDir()
	fp := filepath.Join(base, taskLogFile)
	file, err := os.Open(fp)
	if err != nil {
		return nil, fmt.Errorf("could not open codex_task_tracker.md: %w", err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			fmt.Fprintf(os.Stderr, "file.Close error: %v\n", err)
		}
	}()

	scanner := bufio.NewScanner(file)
	tasks := make(map[string]struct{})
	for scanner.Scan() {
		fields := strings.Split(scanner.Text(), "|")
		if len(fields) < 5 {
			continue
		}
		name := strings.TrimSpace(fields[1])
		status := strings.TrimSpace(fields[3])
		if status == "✅ Done" {
			tasks[name] = struct{}{}
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("scan codex_task_tracker.md: %w", err)
	}
	return tasks, nil
}

// 🧪 Example Usage:
// err := UpdateTaskTracker("Add nil forecast test", "context", "✅ Done", "Codex", "unit coverage", "Improve Forecast Coverage")
// if err != nil {
// 	log.Fatal(err)
// }

File: backend/internal/context/task_logger.py
"""Utilities for logging Codex tasks in Python.

This mirrors the behavior of task_logger.go for Python tooling and tests."""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path

BASE_DIR = ""
CONTEXT = "backend"
BACKLOG_FILE = "backend/backlog.md"
TASK_LOG_FILE = "codex_task_tracker.md"


def normalize_dashes(text: str) -> str:
    """Replace en-dash/em-dash with hyphen."""
    return re.sub(r"[\u2013\u2014]", "-", text)


def extract_task_title(line: str) -> str | None:
    match = re.search(
        r"(?:codex task:\s*`?([^`]+)`?|task\s+\d+:\s*`?([^`]+)`?)",
        line,
        re.IGNORECASE,
    )
    if not match:
        return None
    title = match.group(1) or match.group(2)
    return normalize_dashes(title.strip()) if title else None


@dataclass
class TaskLogEntry:
    """Represents a single row in codex_task_tracker."""

    task_name: str
    phase: str
    status: str
    layer: str = "-"
    domain: str = "-"
    module: str = "-"
    epic: str = "-"
    feature: str = "-"
    description: str = ""
    test_status: str = "-"


def resolve_base_dir() -> Path:
    """Return directory containing backlog.md and codex_task_tracker.md."""
    if BASE_DIR:
        return Path(BASE_DIR)
    directory = Path.cwd()
    while True:
        if (directory / BACKLOG_FILE).exists():
            return directory
        if directory.parent == directory:
            break
        directory = directory.parent
    return directory


def has_duplicate_task(task_name: str) -> bool:
    """Return True if the task already exists in codex_task_tracker.md."""
    base = resolve_base_dir()
    fp = base / TASK_LOG_FILE
    if not fp.exists():
        return False
    with fp.open("r", encoding="utf-8") as file:
        for line in file:
            fields = [f.strip() for f in line.split("|")]
            if len(fields) < 14:
                continue
            if fields[1] == CONTEXT and fields[2] == task_name:
                return True
    return False


def update_task_tracker(
    entry: TaskLogEntry,
    parent_task_name: str | None = None,
) -> None:
    """Append or update a task entry in codex_task_tracker.md."""
    task_name = entry.task_name
    parent = parent_task_name or ""
    if parent and not task_name.startswith(f"{parent} > "):
        task_name = f"{parent} > {task_name}"
    entry.task_name = task_name

    base = resolve_base_dir()
    fp = base / TASK_LOG_FILE
    now = datetime.now().strftime("%Y-%m-%d")

    row_vals = [
        CONTEXT,
        entry.task_name,
        entry.phase,
        entry.status,
        entry.layer,
        entry.domain,
        entry.module,
        entry.epic,
        entry.feature,
        entry.description,
        entry.test_status,
    ]

    if has_duplicate_task(task_name):
        data = fp.read_text(encoding="utf-8").split("\n")
        for i, line in enumerate(data):
            fields = [f.strip() for f in line.split("|")]
            if len(fields) < 14:
                continue
            if fields[1] == CONTEXT and fields[2] == task_name:
                created = fields[12] if len(fields) > 13 else now
                data[i] = "| " + " | ".join(row_vals + [created, now]) + " |"
                break
        fp.write_text("\n".join(data), encoding="utf-8")
    else:
        row = "| " + " | ".join(row_vals + [now, now]) + " |\n"
        with fp.open("a", encoding="utf-8") as file:
            file.write(row)

    clean_backlog()


def append_subtask(
    parent_task: str,
    subtask_entry: TaskLogEntry,
) -> None:
    """Prefix subtask with parent task and delegate to update_task_tracker."""

    update_task_tracker(subtask_entry, parent_task)


def parse_done_tasks() -> set[str]:
    """Return the set of tasks marked ✅ Done."""
    base = resolve_base_dir()
    fp = base / TASK_LOG_FILE
    tasks: set[str] = set()
    if not fp.exists():
        return tasks
    with fp.open("r", encoding="utf-8") as file:
        for line in file:
            fields = [f.strip() for f in line.split("|")]
            if len(fields) < 14:
                continue
            name = normalize_dashes(fields[2].strip())
            status = fields[4]
            if status == "✅ Done":
                tasks.add(name)
                task_base = name
                if ">" in name:
                    task_base = name.split(">", 1)[0].strip()
                    tasks.add(task_base)
                trimmed = task_base.split("-", 1)[0].strip()
                if trimmed and trimmed != task_base:
                    tasks.add(trimmed)
    return tasks


def clean_backlog() -> None:
    """Remove completed tasks from backend/backlog.md."""
    base = resolve_base_dir()
    done = parse_done_tasks()

    done_normalized: set[str] = set()
    for name in done:
        normalized = normalize_dashes(name)
        done_normalized.add(normalized)
        task_base = normalized
        if ">" in normalized:
            task_base = normalized.split(">", 1)[0].strip()
            done_normalized.add(task_base)
        trimmed = task_base.split("-", 1)[0].strip()
        if trimmed and trimmed != task_base:
            done_normalized.add(trimmed)

    backlog_path = base / BACKLOG_FILE
    if not backlog_path.exists():
        return

    backlog_content = backlog_path.read_text(encoding="utf-8")
    lines = backlog_content.split("\n")
    result: list[str] = []
    skip = False

    for line in lines:
        title = extract_task_title(line)
        if title is not None:
            if title in done_normalized:
                skip = True
                continue
            skip = False
        if skip:
            continue
        result.append(line)

    cleaned = "\n".join(result)
    if backlog_content.endswith("\n") and not cleaned.endswith("\n"):
        cleaned += "\n"

    if cleaned != backlog_content:
        backlog_path.write_text(cleaned, encoding="utf-8")
